### PR TITLE
 [Backport release-23.11] checkpointBuildTools: init

### PR DESCRIPTION
**Pull request by bryango**
## Description of changes

Human-based backport of #279487, #167670 and part of #278777 to `release-23.11`, following instructions from the bot in https://github.com/NixOS/nixpkgs/pull/279487#issuecomment-1883828144.

**Note that the initial PR #167670 is actually, _not_ in 23.11, so for this update to apply consistently, I have to backport the parenst as well,** therefore the subject has become `checkpointBuildTools: init`:
```bash
rev=$(
  # rev of the merge commit from #167670
  curl https://api.github.com/repos/NixOS/nixpkgs/pulls/167670 \
  | jq --raw-output .merge_commit_sha
)
git cherry-pick -x "$rev" --mainline=1

rev=69460711750d4bb4718516fdb155145e922ff610 # fix eval
git cherry-pick -x "$rev"
```
If this is undesired, feel free to close! @roberth 

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by bryango** [comment]
@ofborg build tests.checkpointBuildTools

File: pkgs/build-support/checkpoint-build.nix [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+      ${rsync}/bin/rsync \
+        --checksum --times --atimes --chown=$USER:$USER --chmod=+w \
+        -r ${checkpointArtifacts}/outputs/ .
+      patch -p 1 -i "$sourceDifferencePatchFile"
+      rm "$sourceDifferencePatchFile"
+    '';
+  });
+
+  mkCheckpointedBuild = lib.warn
+    "`mkCheckpointedBuild` is deprecated, use `mkCheckpointBuild` instead!"
```
Comment (roberth) [inline pkgs/build-support/checkpoint-build.nix:88]:
```suggestion
    "`mkCheckpointedBuild` is deprecated; use `mkCheckpointBuild` instead."
```

"," doesn't separate sentences; only clauses. However, ";" can and "." does.

Comment (roberth) [reply pkgs/build-support/checkpoint-build.nix:88]: I actually expected a change to this line, reading about alias warnings, but apparently we're allowed to warn in this attribute. Not sure why, or whether this means we could warn in the aliases file, but great!

**Review by roberth — APPROVED** [review]

File: pkgs/build-support/checkpoint-build.nix (hunk 1 above)
Comment (bryango) [reply pkgs/build-support/checkpoint-build.nix]: Whoops, indeed! I will try to "forward-port" this to master next time, perhaps bundled with some other improvements of this function.

Comment (roberth) [reply pkgs/build-support/checkpoint-build.nix]: It's not a big deal tbh, but I just can't help it.

**Pull request merged**